traitlets/traitlets.py

```diff
@@ -644,6 +648,17 @@ def hold(*a):
                 for v in cache.values():
                     self._notify_trait(*v)
 
+    @contextlib.contextmanager
+    def _allow_write(self):
```

Comment:
I'm sure you can come up with workarounds for both of the problems I described. My point is that doing this with a stateful 'is this currently unlocked?' API is probably not the best way to do this, because it will need workarounds for problems like these. I think more elegant, stateless designs are possible.